### PR TITLE
KAFKA-5647: Use KafkaZkClient in ReassignPartitionsCommand and PreferredReplicaLeaderElectionCommand

### PR DESCRIPTION
**Pull request by omkreddy**
*  Use KafkaZkClient in ReassignPartitionsCommand
*  Use KafkaZkClient in PreferredReplicaLeaderElectionCommand
*  Updated test classes to use new methods
*  All existing tests should pass

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by omkreddy** [comment]
@junrao @ijuma 
1. In previous PRs, I have updated TopicCommand and ConfigCommand. In this PR, I am updating ReassignPartitionsCommand, ReassignPartitionsCommand. Since these changes breaks  existing users code, I am going to add a upgrade note. Do you see any other issues with these changes?

2. I have not updated ConsumerGroupCommand methods. These are deprecated old consumer related methods.

3. Should we replace all the usages of TopicAndPartition with TopicPartition class from ReassignPartitionsCommand.

**Comment by ijuma** [comment]
Thanks for the PR. Since we are in the process of converting some of these tools to AdminClient, maybe we can do that instead? We have a PR from @tombentley for `PreferredReplicaLeaderElectionCommand` and he was also working on one for `ReassignPartitionsCommand `.

File: core/src/main/scala/kafka/server/DynamicConfigManager.scala [hunk 1]
```diff
@@ -18,19 +18,16 @@
...
+import kafka.utils.{Json, Logging}
 import kafka.utils.json.JsonObject
-import kafka.zk.{AdminZkClient, KafkaZkClient}
+import kafka.zk.{KafkaZkClient, AdminZkClient, ConfigEntityChangeNotificationZNode, ConfigEntityChangeNotificationSequenceZNode}
 import org.apache.kafka.common.config.types.Password
 import org.apache.kafka.common.security.scram.ScramMechanism
 import org.apache.kafka.common.utils.Time
 
+import scala.collection.JavaConverters._
+import scala.collection._
```
Comment (tedyu) [inline core/src/main/scala/kafka/server/DynamicConfigManager.scala:31]: Do we need line 28 above ?

**Comment by omkreddy** [comment]
@ijuma  It makes sense to avoid multiple iterations of these classes. We can deprecate and later remove ZK based implementations.  I will leave it to you and @junrao to decide the scope/need of this PR.

**Comment by omkreddy** [comment]
@junrao your thoughts on this?

File: core/src/main/scala/kafka/zk/KafkaZkClient.scala [hunk 2]
```diff
@@ -461,6 +461,21 @@ class KafkaZkClient(zooKeeperClient: ZooKeeperClient, isSecure: Boolean) extends
     topicData.getOrElse(topicPartition, Seq.empty)
   }
 
+  /**
+   * Gets all partitions in the cluster
+   * @return
```
Comment (junrao) [inline core/src/main/scala/kafka/zk/KafkaZkClient.scala]: Could we add the return type?

File: core/src/main/scala/kafka/zk/KafkaZkClient.scala [hunk 3]
```diff
@@ -644,9 +659,10 @@ class KafkaZkClient(zooKeeperClient: ZooKeeperClient, isSecure: Boolean) extends
   /**
    * Creates the partition reassignment znode with the given reassignment.
    * @param reassignment the reassignment to set on the reassignment znode.
-   * @throws KeeperException if there is an error while setting or creating the znode
+   * @throws KeeperException if there is an error while creating the znode
    */
   def createPartitionReassignment(reassignment: Map[TopicPartition, Seq[Int]])  = {
+    createRecursive(AdminZNode.path, throwIfPathExists = false)
```
Comment (junrao) [inline core/src/main/scala/kafka/zk/KafkaZkClient.scala]: Could we just use createRecursive to create the full path?

File: core/src/main/scala/kafka/zk/KafkaZkClient.scala [hunk 4]
```diff
@@ -781,6 +797,19 @@ class KafkaZkClient(zooKeeperClient: ZooKeeperClient, isSecure: Boolean) extends
     retryRequestsUntilConnected(deleteRequests)
   }
 
+  /**
+   * Creates preferred replica election znode with partitions undergoing election
+   * @param partitions
+   * @throws KeeperException if there is an error while creating the znode
+   */
+  def createPreferredReplicaElection(partitions: Set[TopicPartition]): Unit  = {
+    createRecursive(AdminZNode.path, throwIfPathExists = false)
```
Comment (junrao) [inline core/src/main/scala/kafka/zk/KafkaZkClient.scala]: Could we just use createRecursive to create the full path?

**Review by junrao — COMMENTED** [review]
@omkreddy : Thanks for the patch. The plan in KIP-183 is to deprecate the zookeeper option in PreferredReplicaCommand and not remove the zookeeper code. So, it's probably still useful to get this patch in so that we can remove the old ZkClient dependency sooner. A few minor comments below. Also, it's probably better to  replace all the usages of TopicAndPartition with TopicPartition class from ReassignPartitionsCommand.

**Comment by ijuma** [comment]
@junrao, what is your plan for removing `ZkUtils` and `AdminUtils`? Probably not before 2.0, right? So, it seems like it would be removed at the same time as the ZK-based parts of `PreferredReplicaCommand`. We can do this PR, but it doesn't seem like it helps us remove `ZkUtils` faster.

**Comment by junrao** [comment]
@ijuma : Yes, we probably can't remove ZkUtils and AdminUtils any time soon. I am just not sure when we can deprecate the zookeeper option in the tools. We probably don't want to keep using the old ZK utils there.

File: core/src/main/scala/kafka/server/DynamicConfigManager.scala (hunk 1 above)
Comment (omkreddy) [reply core/src/main/scala/kafka/server/DynamicConfigManager.scala:31]: Yes, we need the import

**Comment by omkreddy** [comment]
@junrao  Thanks for the review. Updated the PR with review comments. Pls take a look.

Will update remaining test classes in another PR.

**Comment by omkreddy** [comment]
@junrao pinging for review

**Comment by omkreddy** [comment]
retest this please

**Comment by omkreddy** [comment]
@junrao pinging for review

**Comment by junrao** [comment]
@omkreddy : Thanks for the patch. Sorry for the delay. LGTM.

**Pull request closed without merging**